plugins/autoip.rb

```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
+      else
+        abort "There was an error requesting your IP address, IPAM returned code: #{response.code}, message: #{response.body}"
+      end
+
+      # Get the id of the VLAN to request address
 
-      # Get an IP from our IPAM system
-      uri = options[:add_uri].gsub(/SUBNET|HOSTNAME|USER|APIAPP|APITOKEN/, {
-                                     'SUBNET'   => options[:subnet],
-                                     'HOSTNAME' => options[:hostname],
-                                     'USER'     => username,
-                                     'APIAPP'   => options[:apiapp],
-                                     'APITOKEN' => options[:apitoken],
-                                      } )
+      uri = options[:add_uri]
       uri = URI.escape(uri)
-      uri = URI.parse(uri)
+      uri = URI.parse(uri.concat("subnets/cidr/#{options[:subnet]}"))
       http = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl = true
       request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code =="404"
+        abort "The subnet you requested #{options[:subnet]} can't be found message: #{response.body}"
+      elsif response.code != "200"
+        abort "There was an error while searching for the id of the requested subnet #{response.code}, message: #{response.body}"
+      end
+      subnetId = JSON.parse(response.body)["data"][0]["id"]
+
+      # Get the first free address in requested subnet
+
+      uri = options[:add_uri]
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("subnets/#{subnetId}/first_free/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
       response = http.request(request)
       if response.code != "200"
-        abort "There was an error requesting your IP address, IPAM returned #{response.code}"
+        abort "There was an error while requesting firs free address code: #{response.code}, message: #{response.body}"
       end
-      if response.body =~ /Error: subnet not in IPAM/
-        abort "Error: subnet #{options[:subnet]} not in IPAM"
+      options[:ip] = JSON.parse(response.body)["data"]
+
+      # Commit the new IP and hostname to the database.
+
+      uri = options[:add_uri]
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("addresses/?subnetId=#{subnetId}&ip=#{options[:ip]}&hostname=#{options[:hostname]}&owner=#{username}"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Post.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code != "201"
+        abort "There was an error saving the IP and host to the database, returned code: #{response.code}, message: #{response.body}"
       end
-      options[:ip] = response.body
```

Comment:
It doesn't look like you're returning the IP back to the calling module?  You're just outputting it to the terminal?